collagraph/renderers/pyside_renderer.py

```diff
@@ -0,0 +1,163 @@
+from collections import defaultdict
+import logging
+from typing import Any, Callable
+
+from PySide6 import QtCore, QtWidgets
+from PySide6.QtWidgets import (
+    QBoxLayout,
+    QDialogButtonBox,
+    QFormLayout,
+    QGridLayout,
+    QMainWindow,
+    QSpacerItem,
+    QWidget,
+)
+
+from . import Renderer
+from .pyside import camel_case, dialog, name_to_type, widget, window
+
+
+logger = logging.getLogger(__name__)
+
+
+INSERT_MAPPING = {
+    QWidget: widget.insert,
+    QMainWindow: window.insert,
+    QDialogButtonBox: dialog.insert,
+}
+REMOVE_MAPPING = {
+    QWidget: widget.remove,
+}
+SET_ATTR_MAPPING = {
+    QWidget: widget.set_attribute,
+}
+
+
+# Cache for wrapped types
+WRAPPED_TYPES = {}
+
+LAYOUT = {
+    "Box": QBoxLayout,
+    "Grid": QGridLayout,
+    "Form": QFormLayout,
+}
+
+
+def not_implemented(self, *args, **kwargs):
+    raise NotImplementedError(type(self).__name__)
+
+
+class PySideRenderer(Renderer):
+    """PySide6 renderer."""
+
+    def create_element(self, type_name: str) -> Any:
+        """Create an element for the given type."""
+        # Make sure that an app exists before any widgets
+        # are created. Otherwise we might experience a
+        # hard segfault.
+        QtCore.QCoreApplication.instance() or QtWidgets.QApplication()
+
+        # TODO: how to do spacing / stretch?
+        if type_name in ["Spacing", "Stretch"]:
+            return QSpacerItem(0, 0)
+
+        # Create dynamic subclasses which implement `insert`, `set_attribute`
+        # and `remove` methods.
+        # The generated types are cached in WRAPPED_TYPES so they only have
+        # to be generated once and can be used in equality comparisons
+        if type_name in WRAPPED_TYPES:
+            return WRAPPED_TYPES[type_name]()
+
+        original_type = name_to_type(type_name)
+
+        attrs = {
+            "insert": not_implemented,
+            "remove": not_implemented,
+            "set_attribute": not_implemented,
+        }
+        for insert_class in INSERT_MAPPING:
+            if issubclass(original_type, insert_class):
+                attrs["insert"] = INSERT_MAPPING[insert_class]
```

Comment:
You can use this to sort the mapping by subclass specificity (classes deeper in the hierarchy come first):

```python
INSERT_MAPPING= list(sorted(INSERT_MAPPING.items(), key=lambda x: -len(x[0].__mro__)))
```

Which then allows you to `break` in the for loop, and avoids the situation where you select attributes from the wrong subclass:

```python
        for insert_class, insert_attr in INSERT_MAPPING:
            if issubclass(original_type, insert_class):
                attrs["insert"] = insert_attr
                break
```